aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -39,6 +50,194 @@ fn is_string_type(context: &SafeNativeContext, type_arg: &Type) -> SafeNativeRes
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_READ_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    match ty_args[0] {
+        Type::U128 => {
+            // Extract information from aggregator struct reference.
+            let (id, limit) = aggregator_value_as_u128(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit)?;
+            let value =
+                aggregator.read_most_recent_aggregator_value(aggregator_context.resolver)?;
+            Ok(smallvec![Value::u128(value)])
+        },
+        Type::U64 => {
+            let (id, limit) = aggregator_value_as_u64(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit as u128)?;
+            let value =
+                aggregator.read_most_recent_aggregator_value(aggregator_context.resolver)?;
+            if value > u64::MAX as u128 {
+                return Err(SafeNativeError::Abort {
+                    abort_code: EREAD_OUT_OF_BOUNDS,
+                });
+            }
+            Ok(smallvec![Value::u64(value as u64)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+        }),
+    }
+}
+
+/***************************************************************************************************
+ * native fun snapshot(aggregator: &Aggregator): AggregatorSnapshot<u128>;
+ **************************************************************************************************/
+
+fn native_snapshot(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_SNAPSHOT_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    match ty_args[0] {
+        Type::U128 => {
+            let (aggregator_id, _) = aggregator_value_as_u128(&safely_pop_arg!(args, StructRef))?;
+            let id = aggregator_data.snapshot(&aggregator_id);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+                id as u128
+            )]))])
+        },
+        Type::U64 => {
+            let (aggregator_id, _) = aggregator_value_as_u64(&safely_pop_arg!(args, StructRef))?;
+            let id = aggregator_data.snapshot(&aggregator_id);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+                id
+            )]))])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+        }),
+    }
+}
+
```

Comment:
just to note - you'll need to change the functions below to use aggregator data after aggregators are enabled. (but you can leave that for separate PR)